downstairs/src/lib.rs

```diff
@@ -1361,846 +1361,861 @@ impl ActiveConnection {
-            /*
-             * Don't wait more than TIMEOUT_SECS * TIMEOUT_LIMIT seconds to hear
-             * from the other side.
-             * XXX Timeouts, timeouts: always wrong!  Some too short and
-             * some too long.
-             */
-            _ = verbose_timeout(TIMEOUT_SECS, TIMEOUT_LIMIT, log.clone()) => {
-                warn!(log, "inactivity timeout");
-                send_disconnect();
-                break;
+                cdt::work__done!(|| new_id.0);
             }
+        }
+        Ok(())
+    }
 
-            // Check whether the reply task has stopped of its own accord
-            _ = &mut reply_task => {
-                warn!(log, "reply_task stopped");
+    /// Given a job ID (and its associated connection), do the work for that IO
+    ///
+    /// Take a IOop type and (after some error checking), do the work required
+    /// for that IOop, storing the result. On completion, construct the
+    /// corresponding Crucible Message containing the response to it.  The
+    /// caller is responsible for sending that response back to the upstairs.
+    ///
+    /// This function is nominally infallible, because any errors are encoded as
+    /// an error field in the returned `Message`.
+    async fn do_work(
+        &mut self,
+        job_id: JobId,
+        flags: &DownstairsFlags,
+        reqwest_client: &reqwest::Client,
+        region: &mut Region,
+    ) -> Message {
+        let job = self.work.get_ready_job(job_id);
+        let upstairs_connection = self.upstairs_connection;
+        assert_eq!(job.ds_id, job_id);
 
-                // It may be theoretically possible for `reply_task` to stop
-                // (when `ConnectionState::reply_channel_tx` is dropped) and
-                // then for this `select!` to notice it stops **before**
-                // noticing that `cancel_io.cancelled()` is true.
-                //
-                // If that's the case, it would be bad to tell the `Downstairs`
-                // that we've disconnected, because it already knows!
-                //
-                // However, the cancel token is always dropped _before_
-                // `reply_channel_tx` in `ConnectionState`, so we can check
-                // that case here:
-                if !cancel_io.is_cancelled() {
-                    send_disconnect();
+        match job.work {
+            IOop::Read {
+                dependencies,
+                requests,
+            } => {
+                /*
+                 * Any error from an IO should be intercepted here and passed
+                 * back to the upstairs.
+                 */
+                let response = if flags.read_errors && random() && random() {
+                    warn!(self.log, "returning error on read!");
+                    Err(CrucibleError::GenericError("test error".to_string()))
+                } else {
+                    // This clone shouldn't be too expensive, since it's only
```

Comment:
I wonder what clone this comment referred to, as I don't see any clones here.

Reply:
Outdated, good catch!